Add arc.lib: library for substitution ideals from arcs to compute general jet schemes

unepetitemotif opened the pull request:
This PR adds a new Singular library `arc.lib` which provides procedures
for computing substitution ideals from arcs of fat points.

- Main procedure: `arc` computes substitution ideals.
- Supports truncation via integer lists.
- Example usage:

ring R = 0,(x,y),dp;
ideal I = x^4 + y^3;
ideal J = x^2,y^2;
def s = arc(I, J);
setring s;
ideal A = arcideal;
A;

ring R = 0,(x,y),dp;
ideal I = x^4 + y^3;
ideal J = x^2,y^2;
list L=1,2;
def s = arc(I, J, L);
setring s;
ideal A = arcideal;
A;

infinity-groupoids committed:
Add arc.lib: library for substitution ideals from arcs

hannes14 commented:
Thank you for contributing to Singular. please put the example from the commit message also into an example section of
the main procedure.
For the description of the procedure in the library; strings may continue over several lines, but can not conatin \n etc.

unepetitemotif committed:
Update arc.lib

Updated code to remove instances of /n and to add examples found in comment. 

There are more examples which can be added in the GitHub repository under my other GitHub account:

https://github.com/infinity-groupoids/arc.lib

unepetitemotif commented:
Thank you for your comment. I have just noticed that **Sage** has a totally different function called "Arc" which may be producing the failure of the workflow tests. I'd prefer the main procedure to remain with the name arc, but I can alternatively investigate another name if that is necessary. 

Thank you again!

unepetitemotif commented:
The CI failed due to 409 Conflict: artifact already exists. This looks like a workflow config issue rather than code.

The Fedora ARM64 job failed with exit code 2. Other jobs passed. From the logs it looks like an environment/CI issue rather than code.

unepetitemotif commented:
The CI failed with denied: installation not allowed to Create organization package.
This looks like a permissions issue with GitHub Container Registry, not related to my code changes.

unepetitemotif commented:
> Thank you for contributing to Singular. please put the example from the commit message also into an example section of the main procedure. For the description of the procedure in the library; strings may continue over several lines, but can not conatin \n etc.

Can we rerun the checks this week? I believe I updated the file and other checks are CI issues.

unepetitemotif commented:
> Thank you for contributing to Singular. please put the example from the commit message also into an example section of the main procedure. For the description of the procedure in the library; strings may continue over several lines, but can not conatin \n etc.

Hey Hans, I am curious what the next step should be for getting this included as a Singular library for other future users?

hannes14 commented on Singular/LIB/arc.lib:
```diff
@@ -0,0 +1,571 @@
+        return( arc_internal(I, J, optList, useOptList_int) );
+    }
+    print("Usage: arc(I,J) or arc(I,J,optList)");
+    return;
```
This return is not needed, can be removed

hannes14 commented on Singular/LIB/arc.lib:
```diff
@@ -0,0 +1,571 @@
+    }
+    print("Usage: arc(I,J) or arc(I,J,optList)");
+    return;
+}
```
Here an example part is missing:
`example
{ "EXAMPLE:"; echo = 2;
....
}'

hannes14 commented on Singular/LIB/arc.lib:
```diff
@@ -0,0 +1,571 @@
+RETURN:
+  A ring containing the ideal arcideal in the variables of T (minus the original variables in Vars).
+
+EXAMPLE:
```
The example should be in an example part of the (main) procedure, not here

hannes14 commented on Singular/LIB/arc.lib:
```diff
@@ -0,0 +1,571 @@
+    }
+
+    // Create helper ring with identical names, then import
+    string command1 = sprintf("ring Rhelp = 0, (%s), dp;", ambientVarsCSV);
```
```suggestion
    string command1 = "ring Rhelp = 0, ("+ambientVarsCSV+"), dp;";
```

hannes14 commented on Singular/LIB/arc.lib:
```diff
@@ -0,0 +1,571 @@
+    ideal Jhelp = imap(sourceRing, J);
+    
+    // Check fat point is not trivial
+    if (Jhelp == 0) { print("Error: The ideal Jhelp is the zero ideal. Terminating."); return; }
```
```suggestion
    if (Jhelp == 0) { ERROR("The ideal Jhelp is the zero ideal. Terminating.");  }
```

hannes14 commented on Singular/LIB/arc.lib:
```diff
@@ -0,0 +1,571 @@
+    for (int fatVarIndex = 2; fatVarIndex <= size(fatVars); fatVarIndex++) {
+        varsString = varsString + "," + string(fatVars[fatVarIndex]);
+    }
+    string command2 = sprintf("ring F = 0, (%s), dp;", varsString);
```
```suggestion
    string command2 = "ring F = 0, ("+varsString+"), dp;";
```

hannes14 commented on Singular/LIB/arc.lib:
```diff
@@ -0,0 +1,571 @@
+    }
+    string command2 = sprintf("ring F = 0, (%s), dp;", varsString);
+    execute(command2);
+    execute("setring F;");
```
```suggestion
    setring F;
```

hannes14 commented on Singular/LIB/arc.lib:
```diff
@@ -0,0 +1,571 @@
+    // Step 8: new ring T
+    ///////////
+    int totalCoeffCount = numbVars * len_JF;
+    if (totalCoeffCount <= 0) { print("Error: fat point length < 2."); return; }
```
```suggestion
    if (totalCoeffCount <= 0) { ERROR("fat point length < 2.");  }
```

hannes14 commented on Singular/LIB/arc.lib:
```diff
@@ -0,0 +1,571 @@
+    for (int coeffIndex = 1; coeffIndex <= totalCoeffCount; coeffIndex++) {
+        T_vars = T_vars + ",a" + string(coeffIndex);
+    }
+    string command3 = sprintf("ring T = 0, (%s), dp;", T_vars);
```
```suggestion
    string command3 = "ring T = 0, ("+T_vars+"), dp;";
```

hannes14 commented on Singular/LIB/arc.lib:
```diff
@@ -0,0 +1,571 @@
+    }
+    string command3 = sprintf("ring T = 0, (%s), dp;", T_vars);
+    execute(command3);
+    execute("setring T;");
```
```suggestion
    setring T;
```

hannes14 commented on Singular/LIB/arc.lib:
```diff
@@ -0,0 +1,571 @@
+    ideal basis_JF_T = imap(F, basis_JF_F);
+    if (basis_JF_T[len_JF] != 1)
+    {
+        print("Error: last basis_JF_T != 1."); return;
```
```suggestion
       ERROR("last basis_JF_T != 1.");
```

hannes14 commented on Singular/LIB/arc.lib:
```diff
@@ -0,0 +1,571 @@
+        {
+            mapStr = mapStr + ",";
+        }
+    }
```
It is better without execute and string:
```
int substIndex;ideal II;
for(subsIndex=1;substIndex<=numbVars;substIndex++)
{
   II[substIndex]=arcs[substIndex};
}
map substMat=T,II
ideal bigIdeal=substMap(I_J);
```

hannes14 commented on Singular/LIB/arc.lib:
```diff
@@ -0,0 +1,571 @@
+        {
+            
+            for (originalVarScanIndex = 1; originalVarScanIndex <= numbVars; originalVarScanIndex++) {
+                if (deg(currentPoly, originalVarScanIndex) > 0) {
```
deg with 2 arguments expects deg(<poly>,<intvec>).
Here originalVarScanIndex (int) will be converted to intvec(originalVarScanIndex).
Maybe it should be
```
intvec ww;
for (...)
{
  ww=0:vars(basering);
  ww[originalVarScanIndex]=1;
  if(deg(currentPoly,ww)>0) ....
```

hannes14 commented on Singular/LIB/arc.lib:
```diff
@@ -0,0 +1,571 @@
+    }
+    // output a ring with arcideal stored
+    setring sourceRing;
+    string command4 = sprintf("ring S_coeff = 0, (%s), dp;", coeff_vars);
```
```suggestion
    string command4 = "ring S_coeff = 0, ("+coeff_vars+"), dp;";
```

hannes14 commented on Singular/LIB/arc.lib:
```diff
@@ -0,0 +1,571 @@
+    setring sourceRing;
+    string command4 = sprintf("ring S_coeff = 0, (%s), dp;", coeff_vars);
+    execute(command4);
+    string rCmd = sprintf("setring %s;", "S_coeff");
```
```suggestion
   setring S_coeff;
```

hannes14 commented on Singular/LIB/arc.lib:
```diff
@@ -0,0 +1,571 @@
+    string command4 = sprintf("ring S_coeff = 0, (%s), dp;", coeff_vars);
+    execute(command4);
+    string rCmd = sprintf("setring %s;", "S_coeff");
+    execute(rCmd);
```
```suggestion
```

hannes14 commented on Singular/LIB/arc.lib:
```diff
@@ -0,0 +1,571 @@
+    execute(rCmd);
+    string cmdArcIdeal = sprintf("ideal arcideal = %s;", arcideal_str);
+    execute(cmdArcIdeal);
+    string finCmd = sprintf("export %s;", "arcideal");
```
```suggestion
   export arcideal;
```

hannes14 commented on Singular/LIB/arc.lib:
```diff
@@ -0,0 +1,571 @@
+    string cmdArcIdeal = sprintf("ideal arcideal = %s;", arcideal_str);
+    execute(cmdArcIdeal);
+    string finCmd = sprintf("export %s;", "arcideal");
+    execute(finCmd);
```
```suggestion
```

hannes14 commented:
Sorry for the delay, I was not in KL the last 3 weeks.
For the example section of the main procedure see https://www.singular.uni-kl.de/Manual/4-4/sing_54.htm
It would also be nice to have a test file, see https://github.com/Singular/Singular/blob/spielwiese/Tst/README

unepetitemotif committed:
Update arc.lib

implemented all suggested changes by Hans

unepetitemotif committed:
Update arc.lib
